nni/algorithms/hpo/gridsearch_tuner.py

```diff
@@ -2,209 +2,241 @@
+#
+# Here is an example:
+#
+#   search space:
+#     x: choices(5, 7)
+#     y: normal(0, 1)
+#     z: quniform(2, 3, 1)
+#
+#   grid of first epoch:
+#     x: [5, 7]
+#     y: [1/2]
+#     z: [1/2]
+#   generated parameters:
+#     (5,0,3) (7,0,3)
+#
+#   grid of second epoch:
+#     x: [5, 7]
+#     y: [1/2, 1/4, 3/4]  (results in [0, -0.67, 0.67])
+#     z: [1/2, 1/4]  (results in [3, 2], 3/4 is eliminated due to duplication)
+#   generated parameters:
+#     (5,0,2)    (5,-0.67,3) (5,-0.67,2)    (5,0.67,3) (5,0.67,2)
+#     (7,0,2)    (7,-0.67,3) (7,-0.67,2)    (7,0.67,3) (7,0.67,2)
+##
 
 class GridSearchTuner(Tuner):
-    """
-    GridSearchTuner will search all the possible configures that the user define in the searchSpace.
-    The only acceptable types of search space are ``choice``, ``quniform``, ``randint``
+    def __init__(self):
+        self.space = None
+
+        # the grid to search in this epoch
+        # when the space is fully explored, grid is set to None
+        self.grid = None  # list[int | float]
+
+        # a paremter set is internally expressed as a vector
+        # for each dimension i, self.vector[i] is the parameter's index in self.grid[i]
+        # in second epoch of above example, vector [1, 2, 0] means parameters {x: 7, y: 0.67, z: 3}
+        self.vector = None  # list[int]
+
+        # this tells which parameters are derived from previous epoch
+        # in second epoch of above example, epoch_bar is [2, 1, 1]
+        self.epoch_bar = None  # list[int]
+
+        # this stores which intervals are possibly divisible (low < high after log and q)
+        # in first epoch of above example, divisions are:
+        #     {1: [(0,1/2), (1/2,1)], 2: [(0,1/2)]}
+        # in second epoch:
+        #     {1: [(0,1/4), (1/4,1/2), (1/2,3/4), (3/4,1)], 2: [(1/4,1/2)]}
+        # and in third epoch:
+        #     {1: [(0,1/8), ..., (7/8,1)], 2: []}
+        self.divisions = {}  # dict[int, list[tuple[float, float]]]
+
+    def update_search_space(self, space):
+        self.space = format_search_space(space)
+        if not self.space:  # the tuner will crash in this case, report it explicitly
+            raise ValueError('Search space is empty')
+        self._init_grid()
+
+    def generate_parameters(self, *args, **kwargs):
+        params = self._suggest()
+        if params is None:
+            raise nni.NoMoreTrialError('Search space fully explored')
+        return deformat_parameters(params, self.space)
+
+    def receive_trial_result(self, *args, **kwargs):
+        pass
 
-    Type ``choice`` will select one of the options. Note that it can also be nested.
+    def import_data(self, data):  # for resuming
+        # this is not accurate for concurrency > 1
+        # but I believe it's not worthy to introduce complex logic to fix this
+        # instead we should support checkpoint (TODO)
+        for _ in data:
+            self._suggest()
```

Comment:
this is not a correct logic for import data. import_data is not only for resume

Reply:
There is no way to invoke `import_data` except resuming.

Reply:
we provided import data in nnictl, you removed it from nnictl?

Reply:
OK my bad. I have never noticed this command.
But I insist it should be fixed later in another PR.

Reply:
we need to know that this is a broken change

Reply:
It can be fixed in this release. It won't take much time.
I don't want to fix it here because it feels like a new work item. Development of this PR and Yuge's review was finished a week ago so significant logic change should trigger a new round of PR and review.

And, it sounds silly to "import data" to grid search. Don't think [anyone will try it](https://github.com/microsoft/nni/issues?q=is%3Aissue+%22nnictl+experiment+import%22).